fuel-merkle/src/sparse/node.rs

```diff
@@ -195,22 +195,22 @@ impl Node {
     }
 
     pub fn leaf_key(&self) -> &Bytes32 {
-        assert!(self.is_leaf());
+        debug_assert!(self.is_leaf());
         self.bytes_lo()
     }
 
     pub fn leaf_data(&self) -> &Bytes32 {
-        assert!(self.is_leaf());
+        debug_assert!(self.is_leaf());
         self.bytes_hi()
     }
 
     pub fn left_child_key(&self) -> &Bytes32 {
-        assert!(self.is_node());
+        debug_assert!(self.is_node());
         self.bytes_lo()
     }
 
     pub fn right_child_key(&self) -> &Bytes32 {
-        assert!(self.is_node());
+        debug_assert!(self.is_node());
         self.bytes_hi()
     }
```

Comment:
I think public getters should return `Option` instead.

For internal usage, you can create a private version of this function. Maybe even it is better to mark them `unsafe`

Reply:
`Node` is public only to the `fuel-merkle` crate, i.e. `pub(crate) enum Node`. The only place `Node` should ever be used is in the context of the `MerkleTree`, where we already know what kind of node we are using. There is no reason for it to be used in a context where we do not know the kind of node, such as user space or another crate, which is why it is marked `pub(crate)`.  Therefore, I do not agree that these should return `Option` - we always know what data we are accessing, i.e, we could always `unwrap` the `Option`; we would never need to branch based on the variant. 

The `debug_assert`s are there for `fuel-merkle` developers who might modify the update/delete algorithms, not for consumers of the `fuel-merkle` crate. But if you prefer, I can mark these functions as `pub(crate)` as well to make that explicit. 

I also disagree that these functions are unsafe, because it is always "safe" to call them, even in the wrong context. You cannot create an unsafe/incoherent state using these functions.

Reply:
If they are used only within the crate, let's mark them `pub(crate)` to avoid confusion in the future=) 

They are unsafe from a business perspective. I only brought attention to it because [this](https://github.com/FuelLabs/fuel-vm/pull/648/commits/0977a793f369a4a159978b40c5d8b24bc301ea94) commit shows that it is not always true. And from the API perspective, it would be right to have a separate bunch of functions only for internal `MerkleTree` work. And `pub` or `pub(crate)` functions that may fail outside of the `MerkleTree` implementation.

The `unsafe` for internal functions will show that it is dangerous to use. But the comment may describe why it is safe. Or we can always return an option and write `expect` and the reason why it is expected to be valid.

Reply:
> If they are used only within the crate, let's mark them pub(crate) to avoid confusion in the future=)

We can also use `pub(in crate::sparse)` to restrict this scope further. We could even rearrange the file structure to put `Node` under the `merkle_tree` module and do pub(in crate::sparse::merkle_tree)` or really just `pub(super)` to make this very explicit. Any part of the interface marked `pub` is therefore also `pub(in crate::sparse::merkle_tree)` because of Rust's visibility rules.

 > They are unsafe from a business perspective. I only brought attention to it because [this](https://github.com/FuelLabs/fuel-vm/pull/648/commits/0977a793f369a4a159978b40c5d8b24bc301ea94) commit shows that it is not always true. And from the API perspective, it would be right to have a separate bunch of functions only for internal MerkleTree work. And pub or pub(crate) functions that may fail outside of the MerkleTree implementation.
 
I agree that we should differentiate between external and internal APIs, and that they can have different safety guarantees. At the moment, there is really only an internal API for `Node`. Putting `Node` in `Proof` violated that agreement by putting `Node` in user-space and mixing business logic with internal logic. That was bad! `Node` should only be concerned with internal logic, never business logic. Enforcing a more narrow visibility on `Node` will help reduce errors like that in the future.

To summarize: The interface was unsafe from a business perspective, but it was wrong to use it in business logic. I think isolating `Node` to internal logic is the first fix. If you still think `unsafe` should be used in this context, then I will add `unsafe` here.

Reply:
Have a look at https://github.com/FuelLabs/fuel-vm/pull/696. I this is a bit nicer: Here, `Node` is completely restricted to the `merkle_tree` module using `pub(super)`, making it very explicit that the only consumer is `merkle_tree`, and that there are consumers external to `merkle_tree`. This means that the API is strictly internal, and we can always analyze it from an internal perspective.